test/e2e-go/features/transactions/messageRate_test.go

```diff
@@ -0,0 +1,179 @@
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package transactions
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"math"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/test/framework/fixtures"
+)
+
+// this test checks that the txsync outgoing message rate
+// varies according to the transaction rate
+func TestMessageRateChangesWithTxnRate(t *testing.T) {
+	txnRates := []uint{20, 40, 80, 160}
+	if testing.Short() {
+		txnRates = []uint{20, 40}
+	}
+	testMessageRateChangesWithTxnRate(t, filepath.Join("nettemplates", "TwoNodes50Each.json"), txnRates)
+}
+
+func throttleTransactionRate(startTime time.Time, txnRate uint, sentSoFar uint) {
+	timeDelta := time.Since(startTime)
+	currentTps := float64(sentSoFar) / timeDelta.Seconds()
+	if currentTps > float64(txnRate) {
+		sleepDuration := float64(sentSoFar)/float64(txnRate) - timeDelta.Seconds()
+		sleepTime := time.Duration(int64(math.Round(sleepDuration*1000))) * time.Millisecond
+		time.Sleep(sleepTime)
+	}
+}
+
+func testMessageRateChangesWithTxnRate(t *testing.T, templatePath string, txnRates []uint) {
+	t.Parallel()
+	a := require.New(fixtures.SynchronizedTest(t))
+
+	var fixture fixtures.RestClientFixture
+	fixture.SetupNoStart(t, templatePath)
+	configDir := fixture.PrimaryDataDir()
+	cfg, err := config.LoadConfigFromDisk(configDir)
+	a.NoError(err)
+	cfg.EnableVerbosedTransactionSyncLogging = true
+	cfg.SaveToDisk(configDir)
+	fixture.Start()
+
+	defer fixture.Shutdown()
+
+	pingClient := fixture.LibGoalClient
+	pingAccountList, err := fixture.GetWalletsSortedByBalance()
+	a.NoError(err, "fixture should be able to get wallets sorted by balance")
+	pingAccount := pingAccountList[0].Address
+
+	pongClient := fixture.GetLibGoalClientForNamedNode("Node")
+	pongAccounts, err := fixture.GetNodeWalletsSortedByBalance(pongClient.DataDir())
+	a.NoError(err)
+	var pongAccount string
+	for _, acct := range pongAccounts {
+		if acct.Address == pingAccount {
+			continue
+		}
+		// we found an account.
+		pongAccount = acct.Address
+		break
+	}
+
+	a.NotEqual(pingAccount, pongAccount, "accounts under study should be different")
+
+	minTxnFee, minAcctBalance, err := fixture.CurrentMinFeeAndBalance()
+	a.NoError(err)
+
+	transactionFee := minTxnFee + 5
+	amountPingSendsPong := minAcctBalance * 3 / 2
+
+	// build the path for the primary node's log file
+	var logPath strings.Builder
+	logPath.WriteString(fixture.PrimaryDataDir())
+	logPath.WriteString("/node.log")
```

Comment:
avoid assembling paths yourself. Use `filepath.Join(fixture.PrimaryDataDir(), "node.log")` to accomplish that.

Reply:
Fixed.